db/upgrade.php

```diff
@@ -341,5 +341,24 @@ function xmldb_cms_upgrade($oldversion) {
         upgrade_mod_savepoint(true, 2023112100, 'cms');
     }
 
+    if ($oldversion < 2023120101) {
+        // Collect records which have wrong contextid in the customfield data.
+        $sql = "SELECT mcd.id mcdid, mcd.contextid mcdcontextid, mc.id mcid
+                  FROM {customfield_data} mcd
+                  JOIN {customfield_field} mcf ON mcf.id = mcd.fieldid
+                  JOIN {customfield_category} mcc ON mcc.id = mcf.categoryid
+                  JOIN {course_modules} mcm ON mcm.instance = mcd.instanceid AND mcm.module = (
+                      SELECT id FROM mdl_modules WHERE name = 'cms'
+                  )
+                  JOIN {context} mc ON mc.instanceid = mcm.id AND contextlevel = " . CONTEXT_MODULE . "
+                 WHERE mcc.component = 'mod_cms' AND mcd.contextid != mc.id";
+        $records = $DB->get_records_sql($sql);
+        // Update records with correct contextid.
+        foreach ($records as $record) {
+            $DB->set_field('customfield_data', 'contextid', $record->mcid, ['id' => $record->mcdid]);
```

Comment:
I ran proposed SQL in affected prod instance and it brought me 65 710 records. This upgrade can take a while. Maybe shift it to adhoc task?

Reply:
It's moved to the adhoc task now.
I ran in my local and result was like this.
```
Execute adhoc task: mod_cms\task\update_customfield_context
Adhoc task id: 32433048
Adhoc task custom data: 
... started 18:57:43. Current memory use 24.9 MB.
... used 63485 dbqueries
... used 615.87098097801 seconds
Adhoc task complete: mod_cms\task\update_customfield_context
```

Reply:
perfect